Edit HPA In-Context

andrewballantyne opened the pull request:
## Built on #6037 - See the 3rd Commit

**Fixes**: 
https://issues.redhat.com/browse/ODC-4271

**Analysis / Root cause**: 
As a user, I'd like to be able to in-context edit an associated HPA.

**Solution Description**: 
Using the same form from the Add HPA, allow editing.

**Screen shots / Gifs for design review**: 

When a custom metric is in play, we auto select the YAML - they can still use the Form when they so desire to get access to cpu / memory (if applicable), the hpa name and min/max replica counts.
![Screen Shot 2020-07-22 at 7 50 51 AM](https://user-images.githubusercontent.com/8126518/88173176-244f7c80-cbf0-11ea-9b5c-b6183bbe18a7.png)

**Unit test coverage report**: 

```
  hasCustomMetrics accurately determines if an hpa contains non-default metrics
    ✓ expect no metrics to mean no custom metrics
    ✓ expect cpu and memory metrics to not be custom
    ✓ expect any unknown metric to be custom
    ✓ expect an unknown metric to trump known metrics

------------------------------------------------------------|----------|----------|----------|----------|-------------------|
File                                                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
  src/components/hpa/hpa-utils.ts                           |    97.47 |    90.35 |      100 |    97.06 |            92,100 |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
```

**Test setup:**

- Have an HPA on a D/DC Workload
- Using the Kebab option in the Topology sidebar, Edit the HPA

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

andrewballantyne commented:
/hold

Waiting for dependant PR - #6037

christianvogt commented on frontend/packages/dev-console/src/plugin.tsx:
```diff
@@ -808,7 +808,7 @@ const plugin: Plugin<ConsumedExtensions> = [
     properties: {
       exact: true,
-      path: ['/workload-hpa/:ns/:kind/:name'],
+      path: ['/workload-hpa/:ns/:resourceRef/:name', '/workload-hpa/:ns/:resourceRef/:name/edit'],
```
hmmm... Should only have one route. If the HPA already exists, it's an Edit scenario since we have no intention to allow a user to add a second HPA ... ?

andrewballantyne replied:
Ah; didn't think of that. Might be a bit odd if the route is invoked from "Add HPA" and ends up on "Edit HPA" but we don't want to allow them to add two+. I'll code that up.

christianvogt commented on frontend/packages/dev-console/src/components/hpa/HPAPage.tsx:
```diff
@@ -3,20 +3,24 @@ import { Helmet } from 'react-helmet';
     },
   } = props;
-
+  const editHPA = path === '/workload-hpa/:ns/:resourceRef/:name/edit';
```
Derive edit based on presence of HPA?

andrewballantyne commented:
@christianvogt Since it might be difficult to track changes I made here, I'll detail them out:

- I updated the two parts you suggested. Majority of changes are in `HPAPage.tsx` to handle as a single route and just toggle to Edit when an HPA exists
- I updated the Form to disable the name field when in edit, as it'll just prompt an error anyways.

andrewballantyne commented:
See https://github.com/openshift/console/pull/6150
